### PR TITLE
Write temporal delimiters only at the beginning of a packet.

### LINKED ISSUE #954: Temporal Delimiter OBUs are being inserted before every frame
TD OBUs are meant to be present at the beginning of a Temporal Unit, precisely to delimit them, and not before every frame, showable or not, within a given Temporal Unit. All these extra TDs are pointless overhead in the bitstream.

A simple "rav1e INPUT.y4m --output OUTPUT.ivf" seems to be enough to reproduce this. In my tests i used https://0x0.st/sfRh.y4m.

### PR DESCRIPTION
**Pull request by tdaede**
Fixes #954

### CODE REVIEW

File: src/api.rs [hunk 1]
```diff
@@ -339,6 +339,9 @@ impl Config {
       aom_dsp_rtcd();
     }
 
+    let mut packet_data = Vec::new();
+    write_temporal_delimiter(&mut packet_data).unwrap();
```
Comment (negge) [inline src/api.rs]: Please fail gracefully.

File: src/api.rs [hunk 2]
```diff
@@ -593,7 +596,11 @@ impl Context {
 
         let mut fs = FrameState::new(fi);
 
-        let data = encode_frame(fi, &mut fs);
+        let sef_data = encode_frame(fi, &mut fs);
+        self.packet_data.extend(sef_data);
+        let data = self.packet_data.clone();
+        self.packet_data.clear();
+        write_temporal_delimiter(&mut self.packet_data).unwrap();
```
Comment (negge) [inline src/api.rs]: Same thing here.

File: src/api.rs [hunk 3]
```diff
@@ -635,6 +642,7 @@ impl Context {
             if fi.show_frame {
               let data = self.packet_data.clone();
               self.packet_data.clear();
+              write_temporal_delimiter(&mut self.packet_data).unwrap();
```
Comment (negge) [inline src/api.rs]: Ditto.

File: src/api.rs (hunk 1 above)
Comment (tdaede) [reply src/api.rs]: This one is to something I just alloc'd so I know it can't fail, can I leave it?

File: src/encoder.rs [hunk 4]
```diff
@@ -1729,19 +1729,19 @@ fn aom_uleb_encode(mut value: u64, coded_value: &mut [u8]) -> usize {
   leb_size
 }
 
+pub fn write_temporal_delimiter(
+  packet: &mut dyn io::Write
+) -> io::Result<()> {
+  packet.write(&[0x12, 0x00])?;
```
Comment (negge) [inline src/encoder.rs]: Can you use the OBU_Type::OBU_TEMPORAL_DELIMITER enum?

File: src/api.rs [hunk 5]
```diff
@@ -339,6 +339,9 @@ impl Config {
       aom_dsp_rtcd();
     }
 
+    // initialize with temporal delimiter
+    let packet_data = vec![0x12, 0x00];
```
Comment (negge) [inline src/api.rs]: You could hide this inside a constant temporal OBU header somewhere, so you define it once and use it twice.

File: src/encoder.rs (hunk 4 above)
Comment (tdaede) [reply src/encoder.rs]: Yes, but not for the second byte which is the length.

**Pull request merged**